[RFC]fix test_ddp_logging_data_cpu with tsan

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54465 [RFC]fix test_ddp_logging_data_cpu with tsan**

It is reported that there is data race issue when the test runs with tsan. The root cause is from 'model.frc1.double()' call. This is not because DistributedDataParallel() works together with 'model.frc1.double()'. If we remove DistributedDataParallel(), just call 'model.frc1.double(); model.frc2.double();', it complained the same data race issue.

I'm not sure how to do data type cast in this test without tsan complains, so removing this line of codes and mixed data type logging check.

Please kindly let me know if you have a better suggestion on how to do data type cast correctly

Differential Revision: [D27249821](https://our.internmc.facebook.com/intern/diff/D27249821/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27249821/)!

zhaojuanmao committed:
[RFC]fix test_ddp_logging_data_cpu with tsan

It is reported that there is data race issue when the test runs with tsan. The root cause is from 'model.frc1.double()' call. This is not because DistributedDataParallel() works together with 'model.frc1.double()'. If we remove DistributedDataParallel(), just call 'model.frc1.double(); model.frc2.double();', it complained the same data race issue.

I'm not sure how to do data type cast in this test without tsan complains, so removing this line of codes and mixed data type logging check.

Please kindly let me know if you have a better suggestion on how to do data type cast correctly

Differential Revision: [D27249821](https://our.internmc.facebook.com/intern/diff/D27249821/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27249821/)!

[ghstack-poisoned]

rohan-varma reviewed (commented):
If it complains even without DDP and just by calling `model.float() ; model.fc1.double()` then this feels like it's more of an issue with TSAN rather than providing good signal about the test. 

I think we should keep this test to ensure we can log mixed data type models, could we just skip it under TSAN and document why?

zhaojuanmao commented:
> If it complains even without DDP and just by calling `model.float() ; model.fc1.double()` then this feels like it's more of an issue with TSAN rather than providing good signal about the test.
> 
> I think we should keep this test to ensure we can log mixed data type models, could we just skip it under TSAN and document why?

I'm not confident whether TSAN is correct or not...:(.  Your suggestion sounds good to me. Skip TSAN for this test is better than removing mixed data type logging test here. As removing mixed data type logging test did not really resolve TSAN issue anyway.

pritamdamania87 commented:
@zhaojuanmao Could you share the TSAN trace indicating what it is complaining about?

SciPioneer commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -3672,16 +3672,13 @@ def parse_env(var):
             # test larger net with mixed data types, verify multiple bucket sizes
             model = LargeNet()
-            model.float()
-            model.fc1.double()
```
Have you tried `model.float().fc1.double()`?

Out of curiosity, will there still be the same data race issue if change `model.float()` to `model.fc2.float()`?

zhaojuanmao replied:
tried both, Tsan still complained:(

zhaojuanmao commented:
> @zhaojuanmao Could you share the TSAN trace indicating what it is complaining about?

@pritamdamania here is the trace

WARNING: ThreadSanitizer: data race (pid=928094)
  Write of size 1 at 0x7ffc9a684fe0 by main thread:
    #0 memcpy <null> (libtools_build_sanitizers_tsan-py.so+0x10a236)
    #1 at::(anonymous namespace)::empty_strided(c10::ArrayRef<long>, c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>) buck-out/dev/gen/caffe2/aten/gen_aten=RegisterBackendSelect.cpp/RegisterBackendSelect.cpp:128 (libcaffe2_aten_ATen-cpu.so+0x34c01b5)
    #2 operator() caffe2/aten/src/ATen/core/boxing/impl/WrapFunctionIntoFunctor.h:13 (libcaffe2_aten_ATen-cpu.so+0x34bfb93)
    #3 c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (c10::ArrayRef<long>, c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>), &(at::(anonymous namespace)::empty_strided(c10::ArrayRef<long>, c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>))>, at::Tensor, c10::guts::typelist::typelist<c10::ArrayRef<long>, c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool> > >, at::Tensor (c10::ArrayRef<long>, c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>)>::call(c10::OperatorKernel*, c10::DispatchKeySet, c10::ArrayRef<long>, c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>) caffe2/aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:416 (libcaffe2_aten_ATen-cpu.so+0x34bfb93)
    #4 at::Tensor c10::callUnboxedKernelFunction<at::Tensor, c10::ArrayRef<long>, c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool> >(void*, c10::OperatorKernel*, c10::DispatchKeySet, c10::ArrayRef<long>&&, c10::ArrayRef<long>&&, c10::optional<c10::ScalarType>&&, c10::optional<c10::Layout>&&, c10::optional<c10::Device>&&, c10::optional<bool>&&) caffe2/aten/src/ATen/core/boxing/KernelFunction_impl.h:57 (libcaffe2_aten_ATen-cpu.so+0x243fc7b)
    #5 call<at::Tensor, c10::ArrayRef<long>, c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool> > caffe2/aten/src/ATen/core/boxing/KernelFunction_impl.h:67 (libcaffe2_aten_ATen-cpu.so+0x243edf4)
    #6 at::Tensor c10::Dispatcher::callWithDispatchKeySlowPath<at::Tensor, c10::ArrayRef<long>, c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool> >(c10::TypedOperatorHandle<at::Tensor (c10::ArrayRef<long>, c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>)> const&, bool, c10::DispatchKeySet, c10::KernelFunction const&, c10::ArrayRef<long>, c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>) caffe2/aten/src/ATen/core/dispatch/Dispatcher.h:429 (libcaffe2_aten_ATen-cpu.so+0x243edf4)
    #7 call<at::Tensor, c10::ArrayRef<long>, c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool> > caffe2/aten/src/ATen/core/dispatch/Dispatcher.h:452 (libcaffe2_aten_ATen-cpu.so+0x1dad752)
    #8 call caffe2/aten/src/ATen/core/dispatch/Dispatcher.h:390 (libcaffe2_aten_ATen-cpu.so+0x1dad752)
    #9 at::empty_strided(c10::ArrayRef<long>, c10::ArrayRef<long>, c10::TensorOptions) buck-out/dev/gen/caffe2/aten/gen_aten=Functions.cpp/Functions.cpp:3335 (libcaffe2_aten_ATen-cpu.so+0x1dad752)
    #10 at::native::to_impl(at::Tensor const&, c10::TensorOptions const&, bool, bool) caffe2/aten/src/ATen/native/TensorConversions.cpp:38 (libcaffe2_aten_ATen-cpu.so+0x42fd78b)
    #11 at::native::to(at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>) caffe2/aten/src/ATen/native/TensorConversions.cpp:94 (libcaffe2_aten_ATen-cpu.so+0x42fdd59)
    #12 at::(anonymous namespace)::(anonymous namespace)::wrapper_to_dtype(at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>) buck-out/dev/gen/caffe2/aten/gen_aten=RegisterMath.cpp/RegisterMath.cpp:3817 (libcaffe2_aten_ATen-cpu.so+0x39b02cb)
    #13 operator() caffe2/aten/src/ATen/core/boxing/impl/WrapFunctionIntoFunctor.h:13 (libcaffe2_aten_ATen-cpu.so+0x3b88474)
    #14 c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>), &(at::(anonymous namespace)::(anonymous namespace)::wrapper_to_dtype(at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>))>, at::Tensor, c10::guts::typelist::typelist<at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat> > >, at::Tensor (at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>) caffe2/aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:416 (libcaffe2_aten_ATen-cpu.so+0x3b88474)
    #15 at::Tensor c10::callUnboxedKernelFunction<at::Tensor, at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat> >(void*, c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&, c10::ScalarType&&, bool&&, bool&&, c10::optional<c10::MemoryFormat>&&) caffe2/aten/src/ATen/core/boxing/KernelFunction_impl.h:57 (libcaffe2_ATen-core.so+0x4b1f20)
    #16 call<at::Tensor, const at::Tensor &, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat> > caffe2/aten/src/ATen/core/boxing/KernelFunction_impl.h:67 (libcaffe2_ATen-core.so+0x4b1504)
    #17 at::Tensor c10::Dispatcher::callWithDispatchKeySlowPath<at::Tensor, at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat> >(c10::TypedOperatorHandle<at::Tensor (at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>)> const&, bool, c10::DispatchKeySet, c10::KernelFunction const&, at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>) caffe2/aten/src/ATen/core/dispatch/Dispatcher.h:429 (libcaffe2_ATen-core.so+0x4b1504)
    #18 call<at::Tensor, const at::Tensor &, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat> > caffe2/aten/src/ATen/core/dispatch/Dispatcher.h:452 (libcaffe2_ATen-core.so+0x34b765)
    #19 call caffe2/aten/src/ATen/core/dispatch/Dispatcher.h:390 (libcaffe2_ATen-core.so+0x34b765)
    #20 at::Tensor::to(c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>) const buck-out/dev/gen/caffe2/aten/gen_aten=core/TensorMethods.cpp/core/TensorMethods.cpp:3560 (libcaffe2_ATen-core.so+0x34b765)
    #21 torch::autograd::dispatch_to(at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>) buck-out/dev/gen/caffe2/generate-code/autograd/generated/python_variable_methods.cpp:426 (libcaffe2__C_impl_cuda.so+0x2565ccd)
    #22 torch::autograd::THPVariable_to_type(_object*, c10::ScalarType, c10::optional<c10::MemoryFormat>) buck-out/dev/gen/caffe2/generate-code/autograd/generated/python_variable_methods.cpp:537 (libcaffe2__C_impl_cuda.so+0x256475a)
    #23 torch::autograd::THPVariable_double(_object*, _object*, _object*) buck-out/dev/gen/caffe2/generate-code/autograd/generated/python_variable_methods.cpp:588 (libcaffe2__C_impl_cuda.so+0x221bf01)
    #24 method_vectorcall_VARARGS_KEYWORDS /home/engshare/third-party2/python/3.8/src/cpython-3.8//////////Objects/descrobject.c:332 (libpython3.8.so.1.0+0xd49c2)

  Previous read of size 8 at 0x7ffc9a684fe0 by thread T41:
    [failed to restore the stack]

  Location is stack of main thread.

  Location is global '??' at 0x7ffc9a666000 ([stack]+0x00000001efe0)

  Thread T41 (tid=928257, running) created by main thread at:
    #0 pthread_create <null> (libtools_build_sanitizers_tsan-py.so+0x10764d)
    #1 __kmp_create_worker <null> (libomp.so+0x911fa)
    #2 at::TensorIteratorBase::for_each(c10::function_ref<void (char**, long const*, long, long)>, long) caffe2/aten/src/ATen/TensorIterator.cpp:617 (libcaffe2_aten_ATen-cpu.so+0x43324a7)
    #3 void at::TensorIteratorBase::for_each<void at::native::(anonymous namespace)::cpu_kernel<at::native::(anonymous namespace)::copy_kernel(at::TensorIterator&, bool)::$_6::operator()() const::'lambda1'()::operator()() const::'lambda'()::operator()() const::'lambda2'()::operator()() const::'lambda'(float)>(at::TensorIteratorBase&, at::native::(anonymous namespace)::copy_kernel(at::TensorIterator&, bool)::$_6::operator()() const::'lambda1'()::operator()() const::'lambda'()::operator()() const::'lambda2'()::operator()() const::'lambda'(float)&&)::'lambda'(char**, long const*, long), 0>(at::native::(anonymous namespace)::copy_kernel(at::TensorIterator&, bool)::$_6::operator()() const::'lambda1'()::operator()() const::'lambda'()::operator()() const::'lambda2'()::operator()() const::'lambda'(float), long) caffe2/aten/src/ATen/TensorIterator.h:269 (libcaffe2_aten_cpukernel_avx2.so+0x997d0b)
    #4 void at::native::(anonymous namespace)::cpu_kernel<at::native::(anonymous namespace)::copy_kernel(at::TensorIterator&, bool)::$_6::operator()() const::'lambda1'()::operator()() const::'lambda'()::operator()() const::'lambda2'()::operator()() const::'lambda'(float)>(at::TensorIteratorBase&, at::native::(anonymous namespace)::copy_kernel(at::TensorIterator&, bool)::$_6::operator()() const::'lambda1'()::operator()() const::'lambda'()::operator()() const::'lambda2'()::operator()() const::'lambda'(float)&&) caffe2/aten/src/ATen/native/cpu/Loops.h:260 (libcaffe2_aten_cpukernel_avx2.so+0x997b8f)
    #5 at::native::(anonymous namespace)::copy_kernel(at::TensorIterator&, bool)::$_6::operator()() const::'lambda1'()::operator()() const::'lambda'()::operator()() const::'lambda2'()::operator()() const caffe2/aten/src/ATen/native/cpu/CopyKernel.cpp:46 (libcaffe2_aten_cpukernel_avx2.so+0x992eef)
    #6 at::native::(anonymous namespace)::copy_kernel(at::TensorIterator&, bool)::$_6::operator()() const::'lambda1'()::operator()() const::'lambda'()::operator()() const caffe2/aten/src/ATen/native/cpu/CopyKernel.cpp:46 (libcaffe2_aten_cpukernel_avx2.so+0x992a91)
    #7 at::native::(anonymous namespace)::copy_kernel(at::TensorIterator&, bool)::$_6::operator()() const::'lambda1'()::operator()() const caffe2/aten/src/ATen/native/cpu/CopyKernel.cpp:46 (libcaffe2_aten_cpukernel_avx2.so+0x96cf23)
    #8 at::native::(anonymous namespace)::copy_kernel(at::TensorIterator&, bool)::$_6::operator()() const caffe2/aten/src/ATen/native/cpu/CopyKernel.cpp:46 (libcaffe2_aten_cpukernel_avx2.so+0x9420e4)
    #9 at::native::(anonymous namespace)::copy_kernel(at::TensorIterator&, bool) caffe2/aten/src/ATen/native/cpu/CopyKernel.cpp:46 (libcaffe2_aten_cpukernel_avx2.so+0x941112)
    #10 void at::native::DispatchStub<void (*)(at::TensorIterator&, bool), at::native::copy_stub>::operator()<at::TensorIterator&, bool&>(c10::DeviceType, at::TensorIterator&, bool&) caffe2/aten/src/ATen/native/DispatchStub.h:146 (libcaffe2_aten_ATen-cpu.so+0x1a0d924)
    #11 at::native::copy_impl(at::Tensor&, at::Tensor const&, bool) caffe2/aten/src/ATen/native/Copy.cpp:229 (libcaffe2_aten_ATen-cpu.so+0x1a0c521)
    #12 at::native::copy_(at::Tensor&, at::Tensor const&, bool) caffe2/aten/src/ATen/native/Copy.cpp:237 (libcaffe2_aten_ATen-cpu.so+0x1a0ae74)
    #13 at::(anonymous namespace)::(anonymous namespace)::wrapper_copy_(at::Tensor&, at::Tensor const&, bool) buck-out/dev/gen/caffe2/aten/gen_aten=RegisterDefaultBackend.cpp/RegisterDefaultBackend.cpp:391 (libcaffe2_aten_ATen-cpu.so+0x385f8d0)
    #14 operator() caffe2/aten/src/ATen/core/boxing/impl/WrapFunctionIntoFunctor.h:13 (libcaffe2_aten_ATen-cpu.so+0x38b1ee0)
    #15 c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor& (at::Tensor&, at::Tensor const&, bool), &(at::(anonymous namespace)::(anonymous namespace)::wrapper_copy_(at::Tensor&, at::Tensor const&, bool))>, at::Tensor&, c10::guts::typelist::typelist<at::Tensor&, at::Tensor const&, bool> >, at::Tensor& (at::Tensor&, at::Tensor const&, bool)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor&, at::Tensor const&, bool) caffe2/aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:416 (libcaffe2_aten_ATen-cpu.so+0x38b1ee0)
    #16 at::Tensor& c10::callUnboxedKernelFunction<at::Tensor&, at::Tensor&, at::Tensor const&, bool>(void*, c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor&, at::Tensor const&, bool&&) caffe2/aten/src/ATen/core/boxing/KernelFunction_impl.h:57 (libcaffe2_aten_ATen-cpu.so+0x2583419)
    #17 call<at::Tensor &, at::Tensor &, const at::Tensor &, bool> caffe2/aten/src/ATen/core/boxing/KernelFunction_impl.h:67 (libcaffe2_aten_ATen-cpu.so+0x31f9e3e)
    #18 at::Tensor& c10::Dispatcher::redispatch<at::Tensor&, at::Tensor&, at::Tensor const&, bool>(c10::TypedOperatorHandle<at::Tensor& (at::Tensor&, at::Tensor const&, bool)> const&, c10::DispatchKeySet, at::Tensor&, at::Tensor const&, bool) const caffe2/aten/src/ATen/core/dispatch/Dispatcher.h:464 (libcaffe2_aten_ATen-cpu.so+0x31f9e3e)
    #19 redispatch caffe2/aten/src/ATen/core/dispatch/Dispatcher.h:395 (libcaffe2_aten_ATen-cpu.so+0x2ef9611)
    #20 at::redispatch::copy_(c10::DispatchKeySet, at::Tensor&, at::Tensor const&, bool) buck-out/dev/gen/caffe2/aten/gen_aten=RedispatchFunctions.cpp/RedispatchFunctions.cpp:2523 (libcaffe2_aten_ATen-cpu.so+0x2ef9611)
    #21 torch::autograd::VariableType::(anonymous namespace)::copy_(c10::DispatchKeySet, at::Tensor&, at::Tensor const&, bool) caffe2/torch/csrc/autograd/VariableTypeManual.cpp:246 (libcaffe2_libtorch.so+0x1cf6888)
    #22 operator() caffe2/aten/src/ATen/core/boxing/impl/WrapFunctionIntoFunctor.h:13 (libcaffe2_libtorch.so+0x1cf62e5)
    #23 c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor& (c10::DispatchKeySet, at::Tensor&, at::Tensor const&, bool), &(torch::autograd::VariableType::(anonymous namespace)::copy_(c10::DispatchKeySet, at::Tensor&, at::Tensor const&, bool))>, at::Tensor&, c10::guts::typelist::typelist<c10::DispatchKeySet, at::Tensor&, at::Tensor const&, bool> >, at::Tensor& (c10::DispatchKeySet, at::Tensor&, at::Tensor const&, bool)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor&, at::Tensor const&, bool) caffe2/aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:433 (libcaffe2_libtorch.so+0x1cf62e5)
    #24 at::Tensor& c10::callUnboxedKernelFunction<at::Tensor&, at::Tensor&, at::Tensor const&, bool>(void*, c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor&, at::Tensor const&, bool&&) caffe2/aten/src/ATen/core/boxing/KernelFunction_impl.h:57 (libcaffe2_ATen-core.so+0x43b139)
    #25 call<at::Tensor &, at::Tensor &, const at::Tensor &, bool> caffe2/aten/src/ATen/core/boxing/KernelFunction_impl.h:67 (libcaffe2_ATen-core.so+0x43aa05)
    #26 at::Tensor& c10::Dispatcher::callWithDispatchKeySlowPath<at::Tensor&, at::Tensor&, at::Tensor const&, bool>(c10::TypedOperatorHandle<at::Tensor& (at::Tensor&, at::Tensor const&, bool)> const&, bool, c10::DispatchKeySet, c10::KernelFunction const&, at::Tensor&, at::Tensor const&, bool) caffe2/aten/src/ATen/core/dispatch/Dispatcher.h:429 (libcaffe2_ATen-core.so+0x43aa05)
    #27 call<at::Tensor &, at::Tensor &, const at::Tensor &, bool> caffe2/aten/src/ATen/core/dispatch/Dispatcher.h:452 (libcaffe2_ATen-core.so+0x274de3)
    #28 call caffe2/aten/src/ATen/core/dispatch/Dispatcher.h:390 (libcaffe2_ATen-core.so+0x274de3)
    #29 at::Tensor::copy_(at::Tensor const&, bool) const buck-out/dev/gen/caffe2/aten/gen_aten=core/TensorMethods.cpp/core/TensorMethods.cpp:888 (libcaffe2_ATen-core.so+0x274de3)
    #30 at::native::to_impl(at::Tensor const&, c10::TensorOptions const&, bool, bool) caffe2/aten/src/ATen/native/TensorConversions.cpp:41 (libcaffe2_aten_ATen-cpu.so+0x42fd7a9)
    #31 at::native::to(at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>) caffe2/aten/src/ATen/native/TensorConversions.cpp:94 (libcaffe2_aten_ATen-cpu.so+0x42fdd59)
    #32 at::(anonymous namespace)::(anonymous namespace)::wrapper_to_dtype(at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>) buck-out/dev/gen/caffe2/aten/gen_aten=RegisterMath.cpp/RegisterMath.cpp:3817 (libcaffe2_aten_ATen-cpu.so+0x39b02cb)
    #33 operator() caffe2/aten/src/ATen/core/boxing/impl/WrapFunctionIntoFunctor.h:13 (libcaffe2_aten_ATen-cpu.so+0x3b88474)
    #34 c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>), &(at::(anonymous namespace)::(anonymous namespace)::wrapper_to_dtype(at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>))>, at::Tensor, c10::guts::typelist::typelist<at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat> > >, at::Tensor (at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>) caffe2/aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:416 (libcaffe2_aten_ATen-cpu.so+0x3b88474)
    #35 at::Tensor c10::callUnboxedKernelFunction<at::Tensor, at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat> >(void*, c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&, c10::ScalarType&&, bool&&, bool&&, c10::optional<c10::MemoryFormat>&&) caffe2/aten/src/ATen/core/boxing/KernelFunction_impl.h:57 (libcaffe2_ATen-core.so+0x4b1f20)
    #36 call<at::Tensor, const at::Tensor &, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat> > caffe2/aten/src/ATen/core/boxing/KernelFunction_impl.h:67 (libcaffe2_ATen-core.so+0x4b1504)
    #37 at::Tensor c10::Dispatcher::callWithDispatchKeySlowPath<at::Tensor, at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat> >(c10::TypedOperatorHandle<at::Tensor (at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>)> const&, bool, c10::DispatchKeySet, c10::KernelFunction const&, at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>) caffe2/aten/src/ATen/core/dispatch/Dispatcher.h:429 (libcaffe2_ATen-core.so+0x4b1504)
    #38 call<at::Tensor, const at::Tensor &, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat> > caffe2/aten/src/ATen/core/dispatch/Dispatcher.h:452 (libcaffe2_ATen-core.so+0x34b765)
    #39 call caffe2/aten/src/ATen/core/dispatch/Dispatcher.h:390 (libcaffe2_ATen-core.so+0x34b765)
    #40 at::Tensor::to(c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>) const buck-out/dev/gen/caffe2/aten/gen_aten=core/TensorMethods.cpp/core/TensorMethods.cpp:3560 (libcaffe2_ATen-core.so+0x34b765)
    #41 torch::autograd::dispatch_to(at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>) buck-out/dev/gen/caffe2/generate-code/autograd/generated/python_variable_methods.cpp:426 (libcaffe2__C_impl_cuda.so+0x2565ccd)
    #42 torch::autograd::THPVariable_to_type(_object*, c10::ScalarType, c10::optional<c10::MemoryFormat>) buck-out/dev/gen/caffe2/generate-code/autograd/generated/python_variable_methods.cpp:537 (libcaffe2__C_impl_cuda.so+0x256475a)
    #43 torch::autograd::THPVariable_double(_object*, _object*, _object*) buck-out/dev/gen/caffe2/generate-code/autograd/generated/python_variable_methods.cpp:588 (libcaffe2__C_impl_cuda.so+0x221bf01)
    #44 method_vectorcall_VARARGS_KEYWORDS /home/engshare/third-party2/python/3.8/src/cpython-3.8//////////Objects/descrobject.c:332 (libpython3.8.so.1.0+0xd49c2)

SUMMARY: ThreadSanitizer: data race (/data/users/yanlizhao/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_gloo_fork#binary,link-tree/libtools_build_sanitizers_tsan-py.so+0x10a236) in memcpy

zhaojuanmao committed:
Update on "[RFC]fix test_ddp_logging_data_cpu with tsan"


It is reported that there is data race issue when the test runs with tsan. The root cause is from 'model.frc1.double()' call. This is not because DistributedDataParallel() works together with 'model.frc1.double()'. If we remove DistributedDataParallel(), just call 'model.frc1.double(); model.frc2.double();', it complained the same data race issue.

I'm not sure how to do data type cast in this test without tsan complains, so removing this line of codes and mixed data type logging check.

Please kindly let me know if you have a better suggestion on how to do data type cast correctly

Differential Revision: [D27249821](https://our.internmc.facebook.com/intern/diff/D27249821/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27249821/)!

[ghstack-poisoned]

zhaojuanmao commented:
looks like tsan does not work well with fork, so I'm just skipping tsan for all fork tests

wayi1 reviewed (commented):
Also prefer skipping TSAN to deleting some test code.

Once we refactor our test code, we can move all the fork tests to a separate file and disable TSAN on all the tests in that file. @rohan-varma

zhaojuanmao committed:
Update on "[RFC]fix test_ddp_logging_data_cpu with tsan"


It is reported that there is data race issue when the test runs with tsan. The root cause is from 'model.frc1.double()' call. This is not because DistributedDataParallel() works together with 'model.frc1.double()'. If we remove DistributedDataParallel(), just call 'model.frc1.double(); model.frc2.double();', it complained the same data race issue.

I'm not sure how to do data type cast in this test without tsan complains, so removing this line of codes and mixed data type logging check.

Please kindly let me know if you have a better suggestion on how to do data type cast correctly

Differential Revision: [D27249821](https://our.internmc.facebook.com/intern/diff/D27249821/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27249821/)!

[ghstack-poisoned]
